### PR TITLE
prow-controller-manager and sinker: be more strict about build cluster access

### PR DESCRIPTION
**Pull request by listx**
Previously, we only checked if we were able to connect (authenticate
successfully) to build clusters. This change adds an additional check to
see if we can also list pods on the build cluster. This new check will
fail if RBAC roles on the build cluster are not set up correctly.

### CODE REVIEW

**Comment by listx** [comment]
It bugs me a bit that this area of the codebase appears to have little to no testing. The only approach to testing that comes to mind is adding a 2 more KIND clusters, to act as build clusters. One would have correct RBACs configured, while the other would not. Then the code change in this PR should be able to detect that the second build cluster has wrong RBACs and should disable the creation of a "manager" for that build cluster.

File: prow/flagutil/kubernetes_cluster_clients.go
Comment (cjwagner) [inline prow/flagutil/kubernetes_cluster_clients.go:439]: This check isn't sufficient for `prow-controller-manager`. It actually needs permission to create and get pods as well.  (This might be why prow-controller-manager crashlooped when other components that have build cluster clients didn't?)

Rather than actually doing a list/create/get can we instead use the equivalent of `kubectl auth can-i ...`? Listing pods could delay startup significantly and creating a pod would be wasteful and require cleanup.

File: prow/cmd/prow-controller-manager/main.go
Comment (cjwagner) [inline prow/cmd/prow-controller-manager/main.go:170]: :pray: Better yet use a `cluster` log field so it filters properly.

**Comment by chaodaiG** [comment]
> It bugs me a bit that this area of the codebase appears to have little to no testing. The only approach to testing that comes to mind is adding 2 more KIND clusters in our integration tests, to act as build clusters. One would have correct RBACs configured, while the other would not. Then the code change in this PR should be able to detect that the second build cluster has wrong RBACs and should disable the creation of a "manager" for that build cluster.
> 
> Is there some simpler way to do this check in tests?

I don't think you need to create additional clusters. This can be done by generating multiple kubeconfigs for the same build cluster, you might not be able to use gencred directly but it could be helpful

**Review by cjwagner — COMMENTED** [review]
Can you link to the place in the code where the RBAC problem was causing prow-controller-manager to crash? I'm curious as to what was causing this component to crash and not the others that connect to the build clusters.

**Review by chaodaiG — APPROVED** [review]
/hold

**Comment by listx** [comment]
> > It bugs me a bit that this area of the codebase appears to have little to no testing. The only approach to testing that comes to mind is adding 2 more KIND clusters in our integration tests, to act as build clusters. One would have correct RBACs configured, while the other would not. Then the code change in this PR should be able to detect that the second build cluster has wrong RBACs and should disable the creation of a "manager" for that build cluster.
> > Is there some simpler way to do this check in tests?
> 
> I don't think you need to create additional clusters. This can be done by generating multiple kubeconfigs for the same build cluster, you might not be able to use gencred directly but it could be helpful

AFAIK we're not concerned with kubeconfigs (creds for authenticating to a build cluster), but with how the RBAC rules for the build clusters themselves are configured. The scenario that caused the crashloop for us was that we could successfully create a manager [here](https://github.com/kubernetes/test-infra/blob/f1703a819e8570e283123f29a71a8e67aa521f76/prow/flagutil/kubernetes_cluster_clients.go#L371) for the misconfigured build cluster, but that the build manager's RBAC rules were not set up to allow Prow's SA to list pods on it. IOW I think this is an authorization problem, not an authentication problem.

**Comment by listx** [comment]
> Can you link to the place in the code where the RBAC problem was causing prow-controller-manager to crash?

PCM was not crashing, actually (there were no go panics or anything like that). If you look at these logs: https://cloudlogging.app.goo.gl/ni2qDKHLxQ1AwCPP6 you can see that the `fatal` error was `failed to wait for plank caches to sync: timed out waiting for cache to be synced`. This fatal message is from [this line](https://github.com/kubernetes/test-infra/blob/e57429125b2e2617bdd1fc2db00420499e1710a1/prow/cmd/prow-controller-manager/main.go#L206), and it only runs because `mgr.Start(...)` above it stops running. Looking at the [upstream code](https://github.com/kubernetes-sigs/controller-runtime/blob/84c5c9f4e29bfd3cead5c89552c9bc7dc90c45ad/pkg/manager/manager.go#L324-L326) it says that `Start()` can stop when an error occurs, and in our case the error is coming from [here](https://github.com/kubernetes-sigs/controller-runtime/blob/00f2425ce068525e0ff674dba51c3e76ee6ad2da/pkg/source/source.go#L195).

>  I'm curious as to what was causing this component to crash and not the others that connect to the build clusters.

It makes sense that only PCM tripped up here because it's the only component that invokes [Start()](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.1/pkg/manager#RunnableFunc.Start).

File: prow/flagutil/kubernetes_cluster_clients.go
Comment (listx) [reply prow/flagutil/kubernetes_cluster_clients.go:439]: > This check isn't sufficient for `prow-controller-manager`.

True, but it would probably be enough to get PCM past the error that we saw about being unable to list pods.

> It actually needs permission to create and get pods as well.

True --- it would be more robust if we check for those permissions also. I'll compare against the RBAC rules we define for PCM in our integration tests.

> (This might be why prow-controller-manager crashlooped when other components that have build cluster clients didn't?)

Yes, I think so too.

> Rather than actually doing a list/create/get can we instead use the equivalent of `kubectl auth can-i ...`? Listing pods could delay startup significantly and creating a pod would be wasteful and require cleanup.

I didn't know about `kubectl auth can-i`! Definitely sounds like the perfect use case here. I will try to lift that underlying code from upstream and use it here.

**Comment by chaodaiG** [comment]
> > I don't think you need to create additional clusters. This can be done by generating multiple kubeconfigs for the same build cluster, you might not be able to use gencred directly but it could be helpful
> 
> AFAIK we're not concerned with kubeconfigs (creds for authenticating to a build cluster), but with how the RBAC rules for the build clusters themselves are configured. The scenario that caused the crashloop for us was that we could successfully create a manager [here](https://github.com/kubernetes/test-infra/blob/f1703a819e8570e283123f29a71a8e67aa521f76/prow/flagutil/kubernetes_cluster_clients.go#L371) for the misconfigured build cluster, but that the build manager's RBAC rules were not set up to allow Prow's SA to list pods on it. IOW I think this is an authorization problem, not an authentication problem.

I can see that. The way gencred works is creating a cluster SA in the build cluster and authenticating as the service account, gencred attach this SA an admin user RBAC by default, but in your case you can leverage similar flow but not granting the same RBAC to different SA

File: prow/flagutil/kubernetes_cluster_clients.go
Comment (listx) [reply prow/flagutil/kubernetes_cluster_clients.go:439]: FTR, these lines were stolen from https://github.com/kubernetes/test-infra/blob/e57429125b2e2617bdd1fc2db00420499e1710a1/prow/plank/reconciler.go#L230-L233. I suppose we should update that location as well with a `kubectl auth can-i` equivalent? Maybe not in this PR but as a follow-up.

File: prow/flagutil/kubernetes_cluster_clients.go
Comment (listx) [inline prow/flagutil/kubernetes_cluster_clients.go:489]: @cjwagner Does this sound right to you? I'd like a second pair of eyes for this assumption.

Comment (cjwagner) [reply prow/flagutil/kubernetes_cluster_clients.go:489]: > Maybe the right way forward is to just look up the `annotations` field for `iam.gke.io/gcp-service-account` ?

That would work for us, but that is GKE specific.  If we were ok with adding GKE specifics we could just ask `gcloud` what our identity is, but I think it is probably more appropriate to just add a config field to the `plank` config like you suggested in the comment.
That will be provider agnostic, simpler to implement, and would also make this new validation opt-in which is safer at first.

Comment (listx) [reply prow/flagutil/kubernetes_cluster_clients.go:489]: Cool, good to know I was on the right track. Will update this PR to add this in.

Comment (listx) [reply prow/flagutil/kubernetes_cluster_clients.go]: I found a simpler way --- just use SelfSubjectAccessReview instead (this way we essentially defer whatever the `User` should be to the kubeconfig we're using for the build cluster).

**Comment by listx** [comment]
@cjwagner PTAL

File: prow/flagutil/kubernetes_cluster_clients.go
Comment (cjwagner) [inline prow/flagutil/kubernetes_cluster_clients.go:393]: We should check the authorizations without holding the lock so that all the build cluster authorization checks can happen in parallel. I think we only need to hold `lock` when updating `res` or `errs`.

File: prow/flagutil/kubernetes_cluster_clients.go
Comment (cjwagner) [inline prow/flagutil/kubernetes_cluster_clients.go:419]: nit: Lets log this and line 423 at a lower log level (`debug` or `trace`) since this is the expected case. The useful logs are the ones indicating we initially failed to create a build cluster and the log before the `callback()` indicating a build cluster became available. Logging all the broken build clusters from every component every few seconds will be spammy.

File: prow/flagutil/kubernetes_cluster_clients.go
Comment (cjwagner) [inline prow/flagutil/kubernetes_cluster_clients.go:353]: Just a side note for future work in this area:
At some point it might be worth moving the build cluster status tracking logic here rather than determining build cluster status separately. Now that we're checking the authorization here we are already checking for 'reachability' in a more robust way and could return more precise information about what permission is missing for a cluster.
Here is the existing status tracking logic I'm referring to: https://github.com/kubernetes/test-infra/blob/e39fd1ac2edfaf14635705374030734a25415828/prow/plank/reconciler.go#L202-L255

**Review by cjwagner — CHANGES_REQUESTED** [review]
Looks mostly good, I just have a comment about some of the locking and a nit about some logging.

File: prow/flagutil/kubernetes_cluster_clients.go
Comment (listx) [reply prow/flagutil/kubernetes_cluster_clients.go]: Nice catch. I just added more explicit locks/unlocks around here (and also for `res` and `errs`).

File: prow/flagutil/kubernetes_cluster_clients.go
Comment (listx) [reply prow/flagutil/kubernetes_cluster_clients.go:353]: Ack.

File: prow/flagutil/kubernetes_cluster_clients.go
Comment (listx) [reply prow/flagutil/kubernetes_cluster_clients.go]: Done

**Review by listx — COMMENTED** [review]
PTAL

**Review by cjwagner — APPROVED** [review]

File: prow/cmd/prow-controller-manager/main.go
Comment (Multiply) [inline prow/cmd/prow-controller-manager/main.go:168]: I don't believe "get" this is in the starter yaml files, and it took our controller out when I had to force upgrade today due to Github's RSA keys changing.

I simply changed my role to include it, but it might be worth going through all the yaml files and include it, if it's missing.

**Comment by listx** [comment]
> Ack. I think I'll get to this today/tomorrow while the topic is still relatively fresh in my mind.

See https://github.com/kubernetes/test-infra/pull/29232. But basically there was no additional work required. Thanks @cjwagner for the offline analysis/discussion.

File: prow/flagutil/kubernetes_cluster_clients.go
Comment (listx) [reply prow/flagutil/kubernetes_cluster_clients.go:353]: @cjwagner I think I've done this locally (at least a first stab at this). Please have a look at https://github.com/kubernetes/test-infra/pull/29236 to get the ball rolling.

FTR I am trying out a new PR authoring style where I try to create just 1 commit per PR, pretty much like how Gerrit does it because then it (1) results in smaller PRs and (2) allows me to make faster iterative edits because I'm not amending a commit buried somewhere older than HEAD. I got 2 more commits locally but will submit them for review only after we've merged the one above.

**Pull request merged**
